nomad/structs/event.go

```diff
@@ -0,0 +1,155 @@
+package structs
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+
+	multierror "github.com/hashicorp/go-multierror"
+)
+
+// EventStreamRequest is used to stream events from a servers EventBroker
+type EventStreamRequest struct {
+	Topics map[Topic][]string
+	Index  int
+
+	QueryOptions
+}
+
+type EventStreamWrapper struct {
+	Error *RpcError
+	Event *EventJson
+}
+
+type Topic string
+
+const (
+	TopicDeployment Topic = "Deployment"
+	TopicEval       Topic = "Eval"
+	TopicAlloc      Topic = "Alloc"
+	TopicJob        Topic = "Job"
+	TopicNode       Topic = "Node"
+	TopicAll        Topic = "*"
+)
+
+// Event represents a change in Nomads state.
+type Event struct {
+	// Topic represeents the primary object for the event
+	Topic Topic
+
+	// Type is a short string representing the reason for the event
+	Type string
+
+	// Key is the primary identifier of the Event, The involved objects ID
+	Key string
+
+	// Namespace is the namespace of the object, If the object is not namespace
+	// aware (Node) it is left blank
+	Namespace string
+
+	// FilterKeys are a set of additional related keys that are used to include
+	// events during filtering.
+	FilterKeys []string
+
+	// Index is the raft index that corresponds to the event
+	Index uint64
+
+	// Payload is the Event itself see state/events.go for a list of events
+	Payload interface{}
+}
+
+// Events is a wrapper that contains a set of events for a given index.
+type Events struct {
+	Index  uint64
+	Events []Event
+}
+
+// EventJson is a wrapper for a JSON object
+type EventJson struct {
+	Data []byte
+}
+
+func (j *EventJson) Copy() *EventJson {
+	n := new(EventJson)
+	*n = *j
+	n.Data = make([]byte, len(j.Data))
+	copy(n.Data, j.Data)
+	return n
+}
+
+type EventSinkUpsertRequest struct {
+	Sink *EventSink
+	WriteRequest
+}
+
+type EventSinkSpecificRequest struct {
+	ID string
+	QueryOptions
+}
+
+type EventSinkResponse struct {
+	Sink *EventSink
+	QueryMeta
+}
+
+type EventSinkDeleteRequest struct {
+	IDs []string
+	WriteRequest
+}
+
+type EventSinkListRequest struct {
+	QueryOptions
+}
+
+type EventSinkListResponse struct {
+	Sinks []*EventSink
+	QueryMeta
+}
+
+type SinkType string
+
+const (
+	SinkWebhook SinkType = "webhook"
+)
+
+type EventSink struct {
+	ID        string
+	Namespace string
```

Comment:
By my understanding, an `EventSink` collects events from events for both namespaced objects like Jobs and non-namespaced objects like Nodes. What does it mean for an `EventSink` to be namespaced?

Reply:
Yesterday I was thinking that it may be common to setup a type of sink per namespace. The namespace won't affect the Topics a sink is subscribed to, so its only there for namespacing the event sink ID. I think that we can probably remove it though and let users have unique sink IDs.

Reply:
Removed namespacing, I think its not needed, thanks :+1: